src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -274,11 +278,30 @@ public CompiledScript compileInternal(Script script) {
         CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
             //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+            try {
+                compiled = new CompiledScript(script.getType(), name, lang, scriptEngineService.compile(code));
+            } catch (Exception e) {
+                String message = "Compilation error with " + script.getType() + " script ";
+                if (name != null) {
+                    message += "[" + name + "] ";
+                }
+                message += "using [" + lang + "]: " + ExceptionsHelper.detailedMessage(e);
+                throw new ScriptException(message, e);
+            }
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
```

Comment:
@javanna I see three options here:

1. Decide to leave it as is with the benefit being this is the least invasive; however, it is complicated.

2. Add name to the cache key (instead of script type) because there are two cases currently where we can confuse different scripts described in the comment.  The first is if the code is the same between an inline script and an indexed script, you may get back a compiled script with incorrect name and type information.  The second is if two indexed scripts have the same code but different names, you may get back a compiled script with an incorrect name.  Adding name to the cache key would solve both problems.

3. We could leave the cache key as is and cache only the executable script objects instead of compiled script objects.  Then we create a new compiled script object to be returned with the current metadata (name, type, and lang).  I see the point of the cache as trying to prevent compiling the same script over and over, so this solution solves that problem and might actually be the cleanest way of doing this.

Let me know what you think, and then I'll implement the solution we decide upon.  Thanks!